airbyte-cdk/java/airbyte-cdk/db-sources/src/testFixtures/java/io/airbyte/cdk/testutils/ContainerFactory.java

```diff
@@ -22,97 +24,106 @@
  * ContainerFactory is the companion interface to {@link TestDatabase} for providing it with
  * suitable testcontainer instances.
  */
-public interface ContainerFactory<C extends JdbcDatabaseContainer<?>> {
+public abstract class ContainerFactory<C extends JdbcDatabaseContainer<?>> {
 
-  /**
-   * Creates a new, unshared testcontainer instance. This usually wraps the default constructor for
-   * the testcontainer type.
-   */
-  C createNewContainer(DockerImageName imageName);
+  static private final Logger LOGGER = LoggerFactory.getLogger(ContainerFactory.class);
 
-  /**
-   * Returns the class object of the testcontainer.
-   */
-  Class<?> getContainerClass();
+  private record ContainerKey(Class<? extends ContainerFactory> clazz, DockerImageName imageName, List<String> methods) {};
 
-  /**
-   * Returns a shared instance of the testcontainer.
-   */
-  default C shared(String imageName, String... methods) {
-    final String mapKey = Stream.concat(
-        Stream.of(imageName, this.getClass().getCanonicalName()),
-        Stream.of(methods))
-        .collect(Collectors.joining("+"));
-    return Singleton.getOrCreate(mapKey, this);
-  }
+  private static class ContainerOrException {
```

Comment:
Is it worth having a type parameter here? Probably not, just wondering.

Reply:
I thought about it, but it'd be useless because our map will lose the type anyways, so I'll always need to case it.

Reply:
good point, I'd overlooked that